### PR TITLE
Feature [v3] Mention

### PR DESCRIPTION
**Pull request by HoonBaek**
## For Internal Contributors

[UIKIT-1657](https://sendbird.atlassian.net/browse/UIKIT-1657)

## Description Of Changes

Mention
* Add isMentionEnabled props to the <App /> and <SendbirdProvider />
* Add userMention into the config props of the <App /> andd <SendbirdProvider />
  * <SendbirdProvider config={{ userMention: { maxMentionCount: 10, maxSuggestionCount: 15 } }} />
  * maxMentionCount: A maximum count that you can mention in the message input
  * maxSuggestionCount: A maximum user count that the SuggestedMentionList suggests for user mention
* Create SuggestedMentionList component under the Channel smart component
  * Create SuggestedUserMentionItem component
* Create MentionUserLabel ui component
* Add string set
  * MENTION_NAME__NO_NAME: '(No name)'
  * MENTION_COUNT__OVER_LIMIT: 'You can mention up to %d times per message.'

## Types Of Changes

What types of changes does your code introduce to this project?
Put an `x` in the boxes that apply_

- [ ] Bugfix
- [x] New feature
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance (ex) Prettier)
- [ ] Build configuration
- [ ] Improvement (refactor code)

### CODE REVIEW

File: src/smart-components/Channel/components/MessageInput/index.tsx [hunk 1]
```diff
@@ -33,14 +40,49 @@ const MessageInputWrapper = (): JSX.Element => {
   const isOperator = utils.isOperator(channel);
   const { isBroadcast } = channel;
 
+  useEffect(() => {
+    if (mentionedUsers?.length >= 10) {
```
Comment (sravan-s) [inline src/smart-components/Channel/components/MessageInput/index.tsx:44]: what's the reason for `setDisableMention ` when mentionedUsers changes?

Comment (HoonBaek) [reply src/smart-components/Channel/components/MessageInput/index.tsx:44]: It's for mention configuration, we will allow to set up a maximum mentionable count as a configuration

File: .eslintrc.js [hunk 2]
```diff
@@ -38,6 +38,9 @@ module.exports = {
       },
     ],
     "react/forbid-prop-types": 0,
+    // read more - https://github.com/babel/babel-eslint/issues/681#issuecomment-420663038
+    "template-curly-spacing" : "off",
+    indent : "off"
```
Comment (sravan-s) [inline .eslintrc.js:43]: indent : "off" is not a good idea

File: src/lib/Sendbird.jsx [hunk 3]
```diff
@@ -167,6 +169,11 @@ export default function Sendbird(props) {
           pubSub,
           imageCompression,
           useReaction,
+          isMentionEnabled: isMentionEnabled || false,
+          userMention: {
+            maxMentionCount: userMention?.maxMentionCount || 10,
+            maxSuggestionCount: userMention?.maxSuggestionCount || 15,
```
Comment (sravan-s) [inline src/lib/Sendbird.jsx:175]: nit: moving these into some kind of const is cleaner

Comment (HoonBaek) [reply src/lib/Sendbird.jsx:175]: Thanks for the good catch!

File: src/smart-components/Channel/components/SuggestedMentionList/index.tsx [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  onFocusItemChange?: (member: SendBird.User) => void;
+  onFetchUsers?: (users: Array<SendBird.User>) => void;
+  renderUserMentionItem?: (props: { user: SendBird.User }) => JSX.Element;
+  ableAddMention: boolean;
+  maxMentionCount?: number;
+  maxSuggestionCount?: number;
+  inputEvent?: React.KeyboardEvent<HTMLDivElement>;
+}
+
+const DEBOUNCING_TIME = 300;
```
Comment (sravan-s) [inline src/smart-components/Channel/components/SuggestedMentionList/index.tsx:27]: nit: should add this as a prop - can do next release

File: src/smart-components/Channel/components/SuggestedMentionList/index.tsx [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    clearTimeout(timer);
+    setTimer(
+      setTimeout(() => {
+        setSearchString(targetNickname);
+      }, DEBOUNCING_TIME)
+    );
+  }, [targetNickname]);
+
+  useEffect(() => {
+    if (inputEvent?.key === MessageInputKeys.Enter) {
```
Comment (sravan-s) [inline src/smart-components/Channel/components/SuggestedMentionList/index.tsx:64]: 👍

File: src/ui/MessageInput/index.jsx [hunk 6]
```diff
@@ -1,103 +1,330 @@
 import React, {
-  useState,
-  useRef,
-  useEffect,
-  useContext,
+  useState, useRef, useEffect, useCallback, useContext, useMemo,
```
Comment (sravan-s) [inline src/ui/MessageInput/index.jsx:2]: nit: move to perline if more than 3 imports -> I thought the lint rule was there for this

File: src/smart-components/Channel/components/SuggestedMentionList/index.tsx [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    <div
+      className="sendbird-mention-suggest-list"
+      onMouseLeave={() => {
+        if (mouseOverUser) {
+          setCurrentUser(mouseOverUser);
+        }
+      }}
+    >
+      {
+        ableAddMention && currentMemberList.map((member) => (
```
Comment (sravan-s) [inline src/smart-components/Channel/components/SuggestedMentionList/index.tsx:130]: currentMemberList?.map

File: src/smart-components/Channel/components/SuggestedMentionList/index.tsx (hunk 4 above)
Comment (HoonBaek) [reply src/smart-components/Channel/components/SuggestedMentionList/index.tsx:27]: Yes, that's in our plan

**Review by sravan-s — APPROVED** [review]

**Review by sravan-s — APPROVED** [review]

**Pull request merged**